Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx

```diff
@@ -0,0 +1,602 @@
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].getCurrentBuffer().digits.emplace_back(MergerDigit{o2::mch::Digit(deId, padId, adc, time),
+                                                                    stopTime, false, solarId, dsAddr, chAddr});
+}
+
+void Merger::mergeDigits(int feeId)
+{
+  if (feeId < 0 || feeId > MCH_MERGER_FEEID_MAX) {
+    return;
+  }
+
+  mergers[feeId].mergeDigits();
+}
+
+//=======================
+// Data decoder
+void DataDecoder::decodeBuffer(gsl::span<const std::byte> page)
+{
+  size_t ndigits{0};
+
+  uint8_t isStopRDH = 0;
+  uint32_t orbit;
+  uint32_t feeId;
+  uint32_t linkId;
+
+  const auto storeDigit = [&](const Digit& d) {
+    mOutputDigits.emplace_back(d);
+    if (mPrint) {
+      std::cout << "[storeDigit]: digit stored, mOutputDigits size: " << mOutputDigits.size() << std::endl;
+    }
+  };
+
+  auto channelHandler = [&](DsElecId dsElecId, uint8_t channel, o2::mch::raw::SampaCluster sc) {
+    if (mChannelHandler) {
+      mChannelHandler(dsElecId, channel, sc);
+    }
+
+    if (mDs2manu) {
+      channel = ds2manu(int(channel));
+    }
+
+    double digitadc(0);
+    for (auto& s : sc.samples) {
+      digitadc += s;
+    }
+
+    int deId{-1};
+    int dsIddet{-1};
+    if (auto opt = mElec2Det(dsElecId); opt.has_value()) {
+      DsDetId dsDetId = opt.value();
+      dsIddet = dsDetId.dsId();
+      deId = dsDetId.deId();
+    }
+
+    int padId = -1;
+    const Segmentation& segment = segmentation(deId);
+    if ((&segment) == nullptr) {
+      return;
+    }
+
+    padId = segment.findPadByFEE(dsIddet, int(channel));
+    if (mPrint) {
+      auto s = asString(dsElecId);
+      auto ch = fmt::format("{}-CH{:02d}", s, channel);
+      std::cout << ch << "  "
+                << fmt::format("PAD ({:04d} {:04d} {:04d})\tADC {:5.0f}  TIME ({} {} {:02d})  SIZE {}  END {}",
+                               deId, dsIddet, padId, digitadc, orbit, sc.bunchCrossing, sc.sampaTime, sc.nofSamples(), (sc.sampaTime + sc.nofSamples() - 1))
+                << (((sc.sampaTime + sc.nofSamples() - 1) >= 98) ? " *" : "") << std::endl;
+    }
+
+    // skip channels not associated to any pad
+    if (padId < 0) {
+      return;
+    }
+
+    Digit::Time time;
+    time.sampaTime = sc.sampaTime;
+    time.bunchCrossing = sc.bunchCrossing;
+    time.orbit = orbit;
+
+    Digit::Time stopTime;
+    stopTime.sampaTime = sc.sampaTime + sc.nofSamples() - 1;
+    stopTime.bunchCrossing = sc.bunchCrossing;
+    stopTime.orbit = orbit;
+
+    mMerger->addDigit(feeId, static_cast<int>(dsElecId.solarId()), static_cast<int>(dsElecId.elinkId()), static_cast<int>(channel),
+                      deId, padId, digitadc, time, stopTime);
+    ++ndigits;
+  };
+
+  const auto patchPage = [&](gsl::span<const std::byte> rdhBuffer) {
+    auto rdhAnyPtr = reinterpret_cast<o2::header::RDHAny*>(const_cast<std::byte*>(&rdhBuffer[0]));
+    if (mRdhHandler) {
```

Comment:
is this RDHHandler actually used somewhere ?

Reply:
I plan to use it for example in the QC, when directly processing the raw data using the O2 decoder. That's actually the reason why I have introduced it in the first place, such that code using the decoder can still access the sequence of RDHs in the input buffer without having to walk through the data.

What do you think?

Reply:
I always find a bit tricky/risky to introduce code that is not-yet-used (as it's difficult to know beforehand if e.g. it's the right interface, and if/when it will actually be used), but OK, I see your point. Fine with me.